File: api/match-service/src/app/utils/vector_utils.py
import numpy as np
from typing import Dict, List, Tuple, Optional
from math import sqrt, radians, cos, sin, asin
import random

def normalize_vector(vector: np.ndarray) -> np.ndarray:
    """Normalize vector to unit length"""
    norm = np.linalg.norm(vector)
    if norm == 0:
        return vector
    return vector / norm

def cosine_similarity(vector_a: np.ndarray, vector_b: np.ndarray) -> float:
    """
    Calculate cosine similarity between two vectors
    Returns value between -1 and 1, where 1 is most similar
    """
    # Normalize vectors
    norm_a = normalize_vector(vector_a)
    norm_b = normalize_vector(vector_b)
    
    # Calculate cosine similarity
    dot_product = np.dot(norm_a, norm_b)
    
    # Clamp to [-1, 1] to handle floating point errors
    return max(-1.0, min(1.0, dot_product))

def euclidean_distance(vector_a: np.ndarray, vector_b: np.ndarray) -> float:
    """Calculate Euclidean distance between two vectors"""
    return float(np.linalg.norm(vector_a - vector_b))

def weighted_similarity(vector_a: np.ndarray, vector_b: np.ndarray, weights: np.ndarray) -> float:
    """
    Calculate weighted similarity between vectors
    
    Args:
        vector_a: First vector
        vector_b: Second vector  
        weights: Weight vector (same dimensions as vectors)
        
    Returns:
        Weighted similarity score (higher = more similar)
    """
    # Apply weights to the difference
    weighted_diff = weights * (vector_a - vector_b) ** 2
    weighted_distance = sqrt(np.sum(weighted_diff))
    
    # Convert distance to similarity (0-1 scale, higher is better)
    max_possible_distance = sqrt(np.sum(weights))
    if max_possible_distance == 0:
        return 1.0
    
    similarity = 1.0 - (weighted_distance / max_possible_distance)
    return max(0.0, min(1.0, similarity))

def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the great circle distance between two points on Earth
    Returns distance in kilometers
    """
    # Convert decimal degrees to radians
    lat1, lon1, lat2, lon2 = map(radians, [lat1, lon1, lat2, lon2])
    
    # Haversine formula
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a))
    
    # Radius of Earth in kilometers
    r = 6371
    
    return c * r

def user_to_vector(user_data: Dict, include_location: bool = True) -> np.ndarray:
    """
    Convert user data to numerical vector for similarity calculations
    
    Args:
        user_data: User data dictionary (from User.to_matrix_dict())
        include_location: Whether to include lat/lng in vector
        
    Returns:
        NumPy array representing user as vector
    """
    vector_components = []
    
    # Age (normalized to 0-1, assuming age range 18-80)
    age = user_data.get('age', 25)
    vector_components.append((age - 18) / 62)  # Normalize 18-80 to 0-1
    
    # Height (normalized, assuming range 140-220cm)
    height = user_data.get('height', 170)
    vector_components.append((height - 140) / 80)
    
    # Fame (normalized, assuming range 0-100)
    fame = user_data.get('fame', 0)
    vector_components.append(fame / 100)
    
    # Lifestyle attributes (already encoded 0-1)
    lifestyle_attrs = [
        'alcohol_consumption', 'smoking', 'cannabis', 'drugs', 'pets',
        'social_activity_level', 'sport_activity', 'education_level',
        'religion', 'relationship_type', 'children_status',
        'hair_color', 'skin_color', 'eye_color', 'zodiac_sign',
        'gender', 'sex_pref', 'political_view'
    ]
    
    for attr in lifestyle_attrs:
        vector_components.append(user_data.get(attr, 0.0))
    
    # Location (if requested)
    if include_location:
        # Normalize lat/lng to 0-1 ranges
        lat = user_data.get('latitude', 0.0)
        lng = user_data.get('longitude', 0.0)
        
        # Rough normalization (lat: -90 to 90, lng: -180 to 180)
        vector_components.append((lat + 90) / 180)
        vector_components.append((lng + 180) / 360)
    
    return np.array(vector_components, dtype=np.float64)

def create_preference_weights(user_preferences: Optional[Dict] = None) -> np.ndarray:
    """
    Create weight vector for different user attributes
    
    Args:
        user_preferences: User's preference weights
        
    Returns:
        Weight vector matching user_to_vector structure
    """
    # Default weights
    default_weights = {
        'age': 0.2,
        'distance': 0.3,
        'interests': 0.25,  # Covers lifestyle attributes
        'habits': 0.15,     # Covers substance use, etc.
        'relationship': 0.1 # Covers relationship type
    }
    
    if user_preferences:
        default_weights.update(user_preferences)
    
    # Map to vector components
    weights = []
    
    # Age weight
    weights.append(default_weights['age'])
    
    # Height weight (part of physical attributes)
    weights.append(0.1)
    
    # Fame weight (social validation)
    weights.append(0.05)
    
    # Lifestyle attributes (interests/habits)
    lifestyle_weight = default_weights['interests'] / 15  # Distribute among 15 attributes
    for _ in range(15):
        weights.append(lifestyle_weight)
    
    # Location weights (distance)
    weights.extend([default_weights['distance'], default_weights['distance']])
    
    return np.array(weights, dtype=np.float64)

def update_preference_vector(current_vector: np.ndarray, target_vector: np.ndarray, 
                           interaction_type: str, learning_rate: float = 0.1) -> np.ndarray:
    """
    Update user preference vector based on interaction
    
    Args:
        current_vector: Current preference vector
        target_vector: Vector of user being interacted with
        interaction_type: 'like' or 'pass'
        learning_rate: How much to adjust (0-1)
        
    Returns:
        Updated preference vector
    """
    if interaction_type == 'like':
        # Move preference vector closer to liked profile
        direction = target_vector - current_vector
        adjustment = learning_rate * direction
    elif interaction_type == 'pass':
        # Move preference vector away from passed profile
        direction = current_vector - target_vector
        adjustment = learning_rate * direction
    else:
        return current_vector
    
    # Apply adjustment and normalize
    updated_vector = current_vector + adjustment
    return normalize_vector(updated_vector)

def add_randomness(scores: List[Tuple[int, float]], randomness_factor: float = 0.1) -> List[Tuple[int, float]]:
    """
    Add randomness to matching scores to provide variety
    
    Args:
        scores: List of (user_id, score) tuples
        randomness_factor: How much randomness to add (0-1)
        
    Returns:
        Scores with randomness applied
    """
    randomized_scores = []
    
    for user_id, score in scores:
        # Add random adjustment
        random_adjustment = (random.random() - 0.5) * 2 * randomness_factor
        new_score = max(0.0, min(1.0, score + random_adjustment))
        randomized_scores.append((user_id, new_score))
    
    return randomized_scores

def calculate_freshness_boost(days_since_creation: int, max_boost: float = 0.2) -> float:
    """
    Calculate freshness boost for newer profiles
    
    Args:
        days_since_creation: Days since profile was created
        max_boost: Maximum boost to apply
        
    Returns:
        Boost factor (0 to max_boost)
    """
    if days_since_creation <= 7:
        return max_boost * (1 - days_since_creation / 7)
    return 0.0

File: api/match-service/src/utils/vector_utils.go
package utils

import (
	"encoding/json"
	"math"
	"match-service/src/models"
)

// UserVector represents a user as a multi-dimensional vector
type UserVector struct {
	UserID              uint      `json:"user_id"`
	Age                 float64   `json:"age"`                   // Normalized 0-1
	Height              float64   `json:"height"`                // Normalized 0-1
	Fame                float64   `json:"fame"`                  // Normalized 0-1
	AlcoholConsumption  float64   `json:"alcohol_consumption"`   // Encoded 0-1
	Smoking             float64   `json:"smoking"`               // Encoded 0-1
	Cannabis            float64   `json:"cannabis"`              // Encoded 0-1
	Drugs               float64   `json:"drugs"`                 // Encoded 0-1
	Pets                float64   `json:"pets"`                  // Encoded 0-1
	SocialActivityLevel float64   `json:"social_activity_level"` // Encoded 0-1
	SportActivity       float64   `json:"sport_activity"`        // Encoded 0-1
	EducationLevel      float64   `json:"education_level"`       // Encoded 0-1
	Religion            float64   `json:"religion"`              // Encoded 0-1
	ChildrenStatus      float64   `json:"children_status"`       // Encoded 0-1
	PoliticalView       float64   `json:"political_view"`        // Encoded 0-1
	Latitude            float64   `json:"latitude"`              // Normalized 0-1
	Longitude           float64   `json:"longitude"`             // Normalized 0-1
}

// PreferenceVector represents learned user preferences
type PreferenceVector struct {
	UserID      uint        `json:"user_id"`
	Vector      UserVector  `json:"vector"`
	UpdateCount int         `json:"update_count"`
}

// CompatibilityScore represents the result of matching algorithm
type CompatibilityScore struct {
	UserID             uint                   `json:"user_id"`
	CompatibilityScore float64                `json:"compatibility_score"`
	Distance           float64                `json:"distance_km"`
	AgeDifference      int                    `json:"age_difference"`
	Factors            map[string]interface{} `json:"factors"`
}

// MarshalJSON implements custom JSON marshaling for CompatibilityScore
func (c CompatibilityScore) MarshalJSON() ([]byte, error) {
	type Alias CompatibilityScore
	return json.Marshal(&struct {
		*Alias
	}{
		Alias: (*Alias)(&c),
	})
}

// UnmarshalJSON implements custom JSON unmarshaling for CompatibilityScore
func (c *CompatibilityScore) UnmarshalJSON(data []byte) error {
	type Alias CompatibilityScore
	aux := &struct {
		*Alias
	}{
		Alias: (*Alias)(c),
	}
	return json.Unmarshal(data, &aux)
}

// UserToVector converts a User model to a normalized vector
func UserToVector(user *models.User) UserVector {
	vector := UserVector{
		UserID: user.ID,
	}

	// Normalize age (18-80 years)
	if user.Age > 0 {
		vector.Age = math.Max(0, math.Min(1, float64(user.Age-18)/62.0))
	}

	// Normalize height (140-220cm)
	if user.Height.Valid {
		vector.Height = math.Max(0, math.Min(1, float64(user.Height.Int64-140)/80.0))
	}

	// Normalize fame (0-100)
	vector.Fame = math.Max(0, math.Min(1, float64(user.Fame)/100.0))

	// Encode lifestyle attributes
	vector.AlcoholConsumption = encodeLifestyleAttribute(user.AlcoholConsumption.String)
	vector.Smoking = encodeLifestyleAttribute(user.Smoking.String)
	vector.Cannabis = encodeLifestyleAttribute(user.Cannabis.String)
	vector.Drugs = encodeLifestyleAttribute(user.Drugs.String)
	vector.Pets = encodeLifestyleAttribute(user.Pets.String)
	vector.SocialActivityLevel = encodeLifestyleAttribute(user.SocialActivityLevel.String)
	vector.SportActivity = encodeLifestyleAttribute(user.SportActivity.String)
	vector.EducationLevel = encodeLifestyleAttribute(user.EducationLevel.String)
	vector.Religion = encodeLifestyleAttribute(user.Religion.String)
	vector.ChildrenStatus = encodeLifestyleAttribute(user.ChildrenStatus.String)
	vector.PoliticalView = encodeLifestyleAttribute(user.PoliticalView.String)

	// Normalize coordinates (-180 to 180 for longitude, -90 to 90 for latitude)
	if user.Latitude.Valid {
		vector.Latitude = (user.Latitude.Float64 + 90) / 180.0
	}
	if user.Longitude.Valid {
		vector.Longitude = (user.Longitude.Float64 + 180) / 360.0
	}

	return vector
}

// encodeLifestyleAttribute converts lifestyle string to normalized float
func encodeLifestyleAttribute(value string) float64 {
	switch value {
	case "never", "no", "none", "low":
		return 0.0
	case "rarely", "sometimes", "light", "moderate":
		return 0.33
	case "occasionally", "socially", "medium":
		return 0.66
	case "frequently", "regularly", "often", "high", "heavy":
		return 1.0
	default:
		return 0.5 // neutral/unknown
	}
}

// CosineSimilarity calculates cosine similarity between two vectors
func CosineSimilarity(v1, v2 UserVector) float64 {
	dotProduct := 0.0
	magnitude1 := 0.0
	magnitude2 := 0.0

	values1 := vectorToSlice(v1)
	values2 := vectorToSlice(v2)

	for i := 0; i < len(values1); i++ {
		dotProduct += values1[i] * values2[i]
		magnitude1 += values1[i] * values1[i]
		magnitude2 += values2[i] * values2[i]
	}

	if magnitude1 == 0 || magnitude2 == 0 {
		return 0
	}

	return dotProduct / (math.Sqrt(magnitude1) * math.Sqrt(magnitude2))
}

// WeightedSimilarity calculates weighted similarity with custom weights
func WeightedSimilarity(v1, v2 UserVector, weights map[string]float64) float64 {
	similarity := 0.0
	totalWeight := 0.0

	// Define default weights if not provided
	if weights == nil {
		weights = map[string]float64{
			"age":                   0.15,
			"height":                0.05,
			"fame":                  0.10,
			"alcohol_consumption":   0.08,
			"smoking":               0.12,
			"cannabis":              0.06,
			"drugs":                 0.10,
			"pets":                  0.07,
			"social_activity_level": 0.08,
			"sport_activity":        0.08,
			"education_level":       0.06,
			"religion":              0.05,
		}
	}

	// Calculate weighted similarities for each attribute
	attributes := map[string][2]float64{
		"age":                   {v1.Age, v2.Age},
		"height":                {v1.Height, v2.Height},
		"fame":                  {v1.Fame, v2.Fame},
		"alcohol_consumption":   {v1.AlcoholConsumption, v2.AlcoholConsumption},
		"smoking":               {v1.Smoking, v2.Smoking},
		"cannabis":              {v1.Cannabis, v2.Cannabis},
		"drugs":                 {v1.Drugs, v2.Drugs},
		"pets":                  {v1.Pets, v2.Pets},
		"social_activity_level": {v1.SocialActivityLevel, v2.SocialActivityLevel},
		"sport_activity":        {v1.SportActivity, v2.SportActivity},
		"education_level":       {v1.EducationLevel, v2.EducationLevel},
		"religion":              {v1.Religion, v2.Religion},
	}

	for attr, values := range attributes {
		if weight, exists := weights[attr]; exists {
			// Calculate similarity for this attribute (1 - absolute difference)
			attrSimilarity := 1.0 - math.Abs(values[0]-values[1])
			similarity += attrSimilarity * weight
			totalWeight += weight
		}
	}

	if totalWeight > 0 {
		return similarity / totalWeight
	}
	return 0
}

// HaversineDistance calculates geographic distance between two points
func HaversineDistance(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371 // Earth's radius in kilometers

	lat1Rad := lat1 * math.Pi / 180
	lat2Rad := lat2 * math.Pi / 180
	deltaLat := (lat2 - lat1) * math.Pi / 180
	deltaLon := (lon2 - lon1) * math.Pi / 180

	a := math.Sin(deltaLat/2)*math.Sin(deltaLat/2) +
		math.Cos(lat1Rad)*math.Cos(lat2Rad)*
		math.Sin(deltaLon/2)*math.Sin(deltaLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))

	return R * c
}

// UpdatePreferenceVector updates user preferences based on interaction
func UpdatePreferenceVector(preferenceVector, targetVector UserVector, learningRate float64, isPositive bool) UserVector {
	updated := preferenceVector

	values1 := vectorToSlice(preferenceVector)
	values2 := vectorToSlice(targetVector)
	updatedValues := make([]float64, len(values1))

	for i := 0; i < len(values1); i++ {
		if isPositive {
			// Move preference vector closer to target (like)
			updatedValues[i] = values1[i] + learningRate*(values2[i]-values1[i])
		} else {
			// Move preference vector away from target (pass)
			updatedValues[i] = values1[i] - learningRate*(values2[i]-values1[i])
		}
		// Clamp values between 0 and 1
		updatedValues[i] = math.Max(0, math.Min(1, updatedValues[i]))
	}

	return sliceToVector(updatedValues, updated.UserID)
}

// vectorToSlice converts UserVector to slice for calculations
func vectorToSlice(v UserVector) []float64 {
	return []float64{
		v.Age, v.Height, v.Fame,
		v.AlcoholConsumption, v.Smoking, v.Cannabis, v.Drugs, v.Pets,
		v.SocialActivityLevel, v.SportActivity, v.EducationLevel,
		v.Religion, v.ChildrenStatus, v.PoliticalView,
		v.Latitude, v.Longitude,
	}
}

// sliceToVector converts slice back to UserVector
func sliceToVector(values []float64, userID uint) UserVector {
	return UserVector{
		UserID:              userID,
		Age:                 values[0],
		Height:              values[1],
		Fame:                values[2],
		AlcoholConsumption:  values[3],
		Smoking:             values[4],
		Cannabis:            values[5],
		Drugs:               values[6],
		Pets:                values[7],
		SocialActivityLevel: values[8],
		SportActivity:       values[9],
		EducationLevel:      values[10],
		Religion:            values[11],
		ChildrenStatus:      values[12],
		PoliticalView:       values[13],
		Latitude:            values[14],
		Longitude:           values[15],
	}
}